### ISSUE TITLE
Using arena.n.chunk_hooks at initialization 

### ISSUE THREAD

**Issue by mjp41** [jemalloc/jemalloc#700]
I have been trying to use jemalloc where I have intercepted all the calls to VirtualAlloc/mmap.  I have an initialization function that calls `je_mallctl` and updates the chunk hooks:

```C
void Initialize()
{
    unsigned int arenas;
    size_t size = sizeof(unsigned int);
    je_mallctl("arenas.narenas", &arenas, &size, NULL, NULL);

    for(unsigned int i = 0; i < arenas; i++) {
        char buffer[25];
        
        snprintf(buffer, 25, "arena.%d.chunk_hooks", i);

        size_t chunk_struct_size = sizeof(chunk_hooks_t);

        je_mallctl(buffer, NULL, NULL, &mm_hooks, chunk_struct_size);
    }
}
```

But the initial `je_mallctl` call initializes a bunch of chunks before I can replaces the `chunk_hooks`.  So there are some pages that weren't allocated by `mm_hooks`.  Is there a way to either prevent these initial allocations, or discard them from the arenas?

I can guarantee that this code is the first to call into jemalloc.

Thanks

**Comment by jasone** [comment]
The short answer is that no such mechanism currently exists for all arenas.  However, as a side effect of having added support for arena destruction it is now possible in the dev branch for manual (non-automatic) arenas, and if we were to add a mechanism for manually triggered bootstrapping (and disable all uses of ```JEMALLOC_ATTR(constructor)```), the same could be almost trivially done for automatic arenas.  The "arenas.create" mallctl interface takes an optional ```extent_hooks_t *``` argument, and that allows the arena creation code to allocate all metadata via the custom hooks.

If this would be a compelling feature for you, we could develop it as an early test case of our plans for experimental APIs (#567).

**Comment by mjp41** [comment]
This would definitely be interesting for us, and I would be more than happy to be a guinnea pig. 

Just to clarify on our requirements, we only actually require all objects/blocks returned by `je_malloc` and `je_calloc` are on pages from the supplied chunk_hooks. We don't really care where the meta-data is, we need to be able to associate additional meta-data with the return objects.

**Comment by mjp41** [comment]
So perhaps a simpler solution that would address our needs would be to make a build option to make chunk_hooks_default an extern:
```C
extern const chunk_hooks_t     chunk_hooks_default;
```
@jasone Would these be something you would consider?

**Comment by jasone** [comment]
That is probably the right solution, except that we want to mangle the symbol name based on the git hash to make it clear that there are no compatibility guarantees between revisions.  The chunk hooks are particularly worrisome to me from a compatibility standpoint, because changes continue close enough together that I suspect we won't converge on a stable API for some time yet.

**Comment by interwq** [comment]
5.0 added the ability to set the extent hooks when creating arenas, see http://jemalloc.net/jemalloc.3.html#arenas.create

**Comment by ronawho** [comment]
> 5.0 added the ability to set the extent hooks when creating arenas, see http://jemalloc.net/jemalloc.3.html#arenas.create

That doesn't help with hook-replacement for the automatic arenas though, right? Just manually created/managed ones?

**Comment by interwq** [comment]
@ronawho: #1173 will help the case with auto arenas. You can set their extent_hooks at the beginning of the program, even before they get used / initialized. However arena0 is still a special case, and will not behave the same (e.g. its extent hooks cannot allocate since a0 is used to support reentrancy).

**Comment by ronawho** [comment]
Great, thanks for the info!

FWIW we've been sticking with 4.5 because of not being able to replace the extent_hooks for arena 0 in 5.X. 

Currently, we replace the chunk_hooks for all the auto arenas (including arena 0). When running on an HPC system with a NIC that requires registration, we grab a large blob of memory from the system, register it with the NIC, and then satisfy chunk allocations out of that blob. 

Because arena 0 has already grabbed some memory prior to us being able to replace the chunk_hook, we just allocate (and leak) memory for every size class until we see that memory is coming from our blob and that way we can be sure that future allocations will come from the registered blob.

I'm pretty sure I'm abusing the chunk_hooks and that this was never meant to be a supported way of doing things (so it's not surprising that it doesn't work in 5.X), but I had been hopeful for a fix/workaround in 5.1.

**Comment by interwq** [comment]
@ronawho : in the upcoming 5.1 you will be able to replace a0's extent hook, although with some limitation (e.g. allocation in a0's hook won't really work). In fact HHVM does replace a0's extent hook, you can check their use case here: https://github.com/facebook/hhvm/blob/master/hphp/util/alloc.cpp#L417
